### PR TITLE
Better prefix handling for custom annotations

### PR DESCRIPTION
**Pull request by mmdango**
Rather than generating a new `DeepLinkEntry` for each prefix of a custom annotation, we should add the list of prefixes to a single `DeepLinkEntry` so that we only need one entry for each URI of a custom annotation.

For example, say we have a custom annotation, `CustomDeepLink` that specifies the following prefixes:
```
{ "example.com://", "example.co.uk://", "example.fr://", "example.me://" }
```

For an annotation:
```
@CustomDeepLink("foo/bar")
public class SomeActivity
```

The corresponding loader class would have:
```
new DeepLinkEntry("example.com://foo/bar", CLASS, SomeActivity.class, null);
new DeepLinkEntry("example.co.uk://foo/bar", CLASS, SomeActivity.class, null);
new DeepLinkEntry("example.fr://foo/bar", CLASS, SomeActivity.class, null);
new DeepLinkEntry("example.me://foo/bar", CLASS, SomeActivity.class, null);
```

This changes it so that we would instead have:
```
new DeepLinkEntry("foo/bar", CLASS, SomeActivity.class, null, new String[] {"example.com://", "example.co.uk://", "example.fr://", "example.me://"})
```
as a single entry

## Additional work:

I've added some tests as well, let me know if there are any other cases you may have thought of that should be added to this.

Also updated the `AndroidManifest.xml` for the sample project to pick up URIs starting with `library`

@felipecsl

### CODE REVIEW

**Comment by felipecsl** [comment]
Checkstyle failed `/shrug`

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java [hunk 1]
```diff
@@ -25,13 +26,15 @@
   private final DeepLinkEntry.Type annotationType;
   private final TypeElement annotatedElement;
   private final String method;
+  private final String[] prefixes;
+
+  DeepLinkAnnotatedElement(String annotation, Element element, DeepLinkEntry.Type type,
+                           @Nullable String[] prefixes)
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java]: can you pass an empty list instead?

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java [hunk 2]
```diff
@@ -217,22 +220,28 @@ public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment
...
+              asAnnotationValues(AnnotationMirrors.getAnnotationValue(customAnnotation, "value"));
+      String[] prefixesArray = prefixesMap.get(customAnnotation.getAnnotationType().asElement());
+      try {
         for (AnnotationValue suffix : suffixes) {
-          deepLinks.add(prefix + suffix.getValue());
+          deepLinkElements.add(new DeepLinkAnnotatedElement(suffix.getValue().toString(),
+                  element, type, prefixesArray));
         }
+      } catch (MalformedURLException e) {
+        e.printStackTrace();
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: you'll want to throw an annotation processing error here instead of just printing the stack trace

File: deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElementTest.java [hunk 3]
```diff
@@ -21,29 +21,47 @@
...
-        "airbnb://classDeepLink?foo=bar", element, DeepLinkEntry.Type.CLASS);
+        "airbnb://classDeepLink?foo=bar", element, DeepLinkEntry.Type.CLASS, null);
 
     assertThat(annotatedElement.getUri()).isEqualTo("airbnb://classDeepLink?foo=bar");
   }
 
   @Test public void testInvalidUri() {
     try {
-      new DeepLinkAnnotatedElement("http", element, DeepLinkEntry.Type.CLASS);
+      new DeepLinkAnnotatedElement("http", element, DeepLinkEntry.Type.CLASS, null);
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/test/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElementTest.java]: add another constructor to `DeepLinkAnnotatedElement` without the last argument so you dont need to pass it everywhere even when you dont need to

**Comment by mmdango** [comment]
@felipecsl ready for review

File: deeplinkdispatch-processor/doc/deeplinks.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+airbnb://example.com/{foo}/bar\n|#|\nSample doc\n|#|\nDocClass\n|##|\nairbnb://example.com/{foo}/bar\n|#|\n\n|#|\nDocClass#DocMethod\n|##|\n
```
Comment (felipecsl) [inline deeplinkdispatch-processor/doc/deeplinks.txt]: remove this file? also should probably gitignore it

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java [hunk 5]
```diff
@@ -26,13 +27,15 @@
   private final TypeElement annotatedElement;
   private final String method;
   private final Element element;
+  private final String[] prefixes;
+
+  DeepLinkAnnotatedElement(String annotation, Element element, DeepLinkEntry.Type type,
+                           String[] prefixes)
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java]: formatting is off, new lines shouldnt align with the previous

File: sample/src/test/java/com/airbnb/deeplinkdispatch/sample/MainActivityTest.java [hunk 6]
```diff
@@ -98,29 +98,28 @@
...
     assertThat(launchedIntent.getStringExtra("place"), equalTo("testplace"));
     assertThat(launchedIntent.getStringExtra(DeepLink.URI),
-        equalTo("http://example.com/deepLink/testid/testname/testplace"));
+            equalTo("http://example.com/deepLink/testid/testname/testplace"));
 
     Intent parentIntent = shadowApplication.getNextStartedActivity();
     assertNotNull(parentIntent);
     assertThat(parentIntent.getComponent(),
-        equalTo(new ComponentName(deepLinkActivity, MainActivity.class)));
+            equalTo(new ComponentName(deepLinkActivity, MainActivity.class)));
```
Comment (felipecsl) [inline sample/src/test/java/com/airbnb/deeplinkdispatch/sample/MainActivityTest.java]: revert formatting changes?

**Comment by mmdango** [comment]
@felipecsl This just cleans up how prefixes are used by the processor. Instead of making N^2 `DeepLinkEntry`s for prefixes and their URIs, it specifies a list of prefixes for each URI

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java [hunk 7]
```diff
@@ -26,13 +27,14 @@
   private final TypeElement annotatedElement;
   private final String method;
   private final Element element;
+  private final String[] prefixes;
+
+  DeepLinkAnnotatedElement(String annotation, Element element, DeepLinkEntry.Type type,
+          String[] prefixes) throws MalformedURLException {
+    this.prefixes = prefixes;
+
+    validateAnnotation(annotation);
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java]: It's an anti-pattern to throw from a constructor. Either call this method separately after you construct the object or use a builder. While you're at it, I'd also turn `prefixes` into a `List<String>` just because it's easier to work with

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java [hunk 8]
```diff
@@ -46,7 +48,44 @@
...
+    }
+  }
+
+  /**
+   * If the element has prefixes, this will return the URI using the first prefix, otherwise it will
+   * return just {@link DeepLinkAnnotatedElement#uri}
+   * @return the URI represented by this element
+   */
   String getUri() {
+    if (prefixes.length > 0) {
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkAnnotatedElement.java]: Is this method still being called anywhere?

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java [hunk 9]
```diff
@@ -307,12 +322,28 @@ public int compare(DeepLinkAnnotatedElement element1, DeepLinkAnnotatedElement e
...
+      prefixArgumentBuilder
+              .append("\"")
+              .append(prefix)
+              .append("\"");
+
+      if (j < prefixes.length - 1) {
+        prefixArgumentBuilder.append(", ");
+      }
+    }
+    prefixArgumentBuilder.append("}");
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: Another reason to use a `List` so you don't have to do this manual `StringBuilder` stitching

File: sample/src/test/java/com/airbnb/deeplinkdispatch/sample/MainActivityTest.java [hunk 10]
```diff
@@ -97,30 +97,29 @@
...
     assertThat(launchedIntent.getStringExtra("place"), equalTo("testplace"));
     assertThat(launchedIntent.getStringExtra(DeepLink.URI),
-        equalTo("http://example.com/deepLink/testid/testname/testplace"));
+          equalTo("http://example.com/deepLink/testid/testname/testplace"));
 
     Intent parentIntent = shadowApplication.getNextStartedActivity();
     assertNotNull(parentIntent);
-    assertThat(parentIntent.getComponent(),
-        equalTo(new ComponentName(deepLinkActivity, MainActivity.class)));
+    assertThat(parentIntent.getComponent(), equalTo(new ComponentName(deepLinkActivity, MainActivity.class)));
```
Comment (felipecsl) [inline sample/src/test/java/com/airbnb/deeplinkdispatch/sample/MainActivityTest.java]: revert these changes?

File: deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkEntry.java [hunk 11]
```diff
@@ -32,22 +37,50 @@
...
   public enum Type {
     CLASS,
     METHOD
   }
 
-  public DeepLinkEntry(String uri, Type type, Class<?> activityClass, String method) {
-    DeepLinkUri parsedUri = DeepLinkUri.parse(uri);
-    String schemeHostAndPath = schemeHostAndPath(parsedUri);
+  public DeepLinkEntry(String uri, Type type, Class<?> activityClass, String method,
+                       @Nullable String[] prefixes) {
```
Comment (felipecsl) [inline deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkEntry.java]: Use an empty list instead of null

**Comment by rossbacher** [comment]
This might still make sense but for now I'm closing it as it is not compatible with the new matcher.
